### PR TITLE
feat: Use Rack::Events for instrumentation

### PR DESCRIPTION
**Pull request by arielvalentin**
This is an alternate implementation of the Rack instrumentation that leverages the [`Rack::Events` API](https://www.rubydoc.info/gems/rack/Rack/Events) instead of a custom`Rack::Middleware`.

Why am I suggesting we change this instrumentation? At GitHub we leverage `Rack::BodyProxy` to write Rack Request logs after the request is complete, however Rack span is already finished and its related `Context` has already been detached. This means we are not able to correlate request logs to our traces.

The advantage of using `Rack::Events` is that handlers are triggered during difference stages of a request, including for deferred operations like [`Rack::BodyProxy`](https://www.rubydoc.info/gems/rack/Rack/BodyProxy), as opposed to middlewares who _only_ are invoked inline.

The disadvantage of this API is that it makes managing the request more difficult and we have to track the Context positions to detach in the `Rack::Env`.

This implementation will be released along side the existing instrumtation to give users the option to use the middleware instead of the `Rack::Events` handler until we are able to run this in some of our heavy production workloads.

Other noteworthy things:

1. The `http.proxy` span has been converted to a Span Event and ignores the `record_frontend_span` option
2. There are no _new_ events added that we brainstormed in this discussion: https://github.com/open-telemetry/opentelemetry-ruby-contrib/pull/342#discussion_r1113462298
3. All configuration options for this middleware that required special processing/registration were moved into the `Rack::Instrumentation.install` 
4. The Rack event now **includes** the time it takes to write the response to the socket.
5. Sinatra 1.4 depends on Rack 1 so we must support both Rack Middleware and Events unless we choose to drop support for it.

Fixes https://github.com/open-telemetry/opentelemetry-ruby-contrib/issues/341 Related https://github.com/DataDog/dd-trace-rb/pull/1746

### CODE REVIEW

File: instrumentation/rack/lib/opentelemetry/instrumentation/rack/middlewares/event_handler.rb
Comment (fbogsany) [inline instrumentation/rack/lib/opentelemetry/instrumentation/rack/middlewares/event_handler.rb:152]: It might also be nice to add `on_send` and record a span event for "sent first byte" or something similar. It is really nice to know, from a perf analysis perspective, when the application has finalized the response and the remaining work is "out of our hands".

An interesting philosophical question: when should we consider the server span "finished"? When application processing is complete or when the response has been completely written to the socket? I _think_ the current middleware implementation is effectively equivalent to "when application processing is complete", before anything has been written to the socket.

Comment (arielvalentin) [reply instrumentation/rack/lib/opentelemetry/instrumentation/rack/middlewares/event_handler.rb:152]: > It might also be nice to add `on_send` and record a span event for "sent first byte" or something similar. It is really nice to know, from a perf analysis perspective, when the application has finalized the response and the remaining work is "out of our hands".
> 

I like this idea a lot! I will add it to my todo list.

> 
> An interesting philosophical question: when should we consider the server span "finished"? When application processing is complete or when the response has been completely written to the socket? I _think_ the current middleware implementation is effectively equivalent to "when application processing is complete", before anything has been written to the socket.

- 💭 I'm not sure. I lean towards after the middleware has completed writing the response but does that make it asymmetric from when it's processing the incoming request? I'm not entirely certain.

File: instrumentation/rack/lib/opentelemetry/instrumentation/rack/middlewares/event_handler.rb
Comment (arielvalentin) [inline instrumentation/rack/lib/opentelemetry/instrumentation/rack/middlewares/event_handler.rb:61]: @fbogsany @mwear Am I correct in that the frontend span is here for parity with `dd-trace-rb`? Do we have any insight into how popular this feature is? Would it be ok to consider dropping it?

https://github.com/DataDog/dd-trace-rb/blob/2b163dd91c8d349ceb92370c16f4a7be4a691c35/docs/GettingStarted.md#http-request-queuing

I see you all discussed how this should work early: https://github.com/open-telemetry/opentelemetry-specification/issues/51

AFAICT representing proxy spans remains unresolved: https://github.com/open-telemetry/opentelemetry-specification/issues/51

But it looks like we took this approach:
https://github.com/open-telemetry/opentelemetry-specification/pull/337#issuecomment-549567281

One thing I noticed about porting this functionality to use the events API is that the `http.proxy` span does not include semantic attributes although it is the `server` span. Is that intentional? 

https://github.com/open-telemetry/opentelemetry-ruby-contrib/blob/main/instrumentation/rack/lib/opentelemetry/instrumentation/rack/middlewares/tracer_middleware.rb#L109

File: instrumentation/rack/lib/opentelemetry/instrumentation/rack/middlewares/event_handler.rb
Comment (arielvalentin) [inline instrumentation/rack/lib/opentelemetry/instrumentation/rack/middlewares/event_handler.rb:245]: @open-telemetry/ruby-contrib-approvers When I started this implementation I used constructor injection, but then soon realized there are some limitations with that approach that supports _injecting_ this middleware _after_ the Rack app is initialized like in this example: 

https://github.com/open-telemetry/opentelemetry-ruby-contrib/blob/main/instrumentation/rack/example/trace_demonstration2.rb#L27

- Manual instantiation required:  The existing implementation requires users to explicitly add Rack tracer middleware to the builder via `use`. Following the same pattern for the `EventHandler` object instance _may_ be initialized _prior_ to the SDK initialization.  
- Stale config variables: The SDK is lazily initialized so the configuration options may not yet be set on the instrumentation instance. That means the `EventHandler` has to defer looking up the configuration options until _after_ the SDK is initialized, however it does not know when that happens because there aren't any lifecycle hooks that would notify this object instance that the config options are available. Using instance variables made it so the EventHandler had a set of default options and did not work correctly. 
- Tracer not upgraded: The object instance had a stale instance of a tracer, which did not make any sense to me as I expected the tracer to be a upgraded once the SDK was initialized so none of the test passed.

The middleware [instrumentation](https://github.com/open-telemetry/opentelemetry-ruby-contrib/blob/main/instrumentation/rack/lib/opentelemetry/instrumentation/rack/middlewares/tracer_middleware.rb) _memoizes_ these options as class variables. That caught me by surprise since I thought it made more sense for these options to be configured/memoized as part of the instrumentation `instance`. 

Is there a case where Rack requests will start to be processed before the SDK will be initialized?

Does the existing rack middleware account for the instrumentation being disabled via environment variables or the config attributes?

File: instrumentation/rack/lib/opentelemetry/instrumentation/rack/middlewares/event_handler.rb
Comment (arielvalentin) [reply instrumentation/rack/lib/opentelemetry/instrumentation/rack/middlewares/event_handler.rb:61]: We discussed this at the SIG today. 

@robbkidd and I are going to swarm on resolving proxy spans with the HTTP WG but we will be dropping this feature.

Endusers may add `X-Request-Start` or `X-Queue-Time` via `allowed_request_headers` and process the span attributes in their vendor backends.

File: instrumentation/rack/lib/opentelemetry/instrumentation/rack/middlewares/event_handler.rb
Comment (arielvalentin) [inline instrumentation/rack/lib/opentelemetry/instrumentation/rack/middlewares/event_handler.rb:100]: hmm... I should not stop here on the _first_ failure but attempt to process all of the propagators until it has gone through the entire list.

File: instrumentation/rack/lib/opentelemetry/instrumentation/rack/middlewares/event_handler.rb
Comment (arielvalentin) [inline instrumentation/rack/lib/opentelemetry/instrumentation/rack/middlewares/event_handler.rb:135]: hmm... should I add error handling here and attempt to detach each token anyway?

File: instrumentation/rack/lib/opentelemetry/instrumentation/rack/middlewares/event_handler.rb
Comment (fbogsany) [reply instrumentation/rack/lib/opentelemetry/instrumentation/rack/middlewares/event_handler.rb:61]: I don't recall why we added this, but parity with DD seems likely. I would actually prefer to add an Event with the `request_start_time` - that's more meaningful than a "fake span" and is similar to what we do with Sidekiq job instrumentation with `created_at` and `enqueued_at` https://github.com/open-telemetry/opentelemetry-ruby-contrib/blob/57cabe3fac44a0921af542b62dc34ef285104a5c/instrumentation/sidekiq/lib/opentelemetry/instrumentation/sidekiq/middlewares/server/tracer_middleware.rb#L35-L36

In general, I am not fond of encoding timestamps in attributes (as integers or strings) and then parsing them on the backend. This sort of thing is exactly what Events are intended for.

Comment (arielvalentin) [reply instrumentation/rack/lib/opentelemetry/instrumentation/rack/middlewares/event_handler.rb:61]: We did discuss recording it as an event but we didn't know what it would mean for a span event to have a timestamp that was _earlier_ than start timestamp of the event, so I thought it was best not to add it at all.

If users choose to add it to the allow list still that gives them access to the header without us having to add special logic.

Comment (fbogsany) [reply instrumentation/rack/lib/opentelemetry/instrumentation/rack/middlewares/event_handler.rb:61]: There is ample precedent for an event timestamp earlier than the span start time - in particular, see the Sidekiq examples I shared. I strongly believe this should be an event rather than an attribute or a span. Requiring backends to parse and interpret stringly-typed attributes is bad form IMO, especially when we have a well structured and typed alternative.

Comment (ericmustin) [reply instrumentation/rack/lib/opentelemetry/instrumentation/rack/middlewares/event_handler.rb]: Fine to drop and fine to restructure as an event. It was a thing that came from dd and wasn’t even standard across dd sdks, it predated the open tracing Nginx instrumentation iirc and was added as a one off work around for a specific user, although over time it did grow to become a somewhat popular feature with ddtracerb users. I believe the correct approach is to “simply” instrument your proxy.

Comment (arielvalentin) [reply instrumentation/rack/lib/opentelemetry/instrumentation/rack/middlewares/event_handler.rb]: I have moved it into an event.

File: instrumentation/rack/lib/opentelemetry/instrumentation/rack/middlewares/event_handler.rb
Comment (arielvalentin) [reply instrumentation/rack/lib/opentelemetry/instrumentation/rack/middlewares/event_handler.rb:213]: I moved all of the configuration setup into the `install` block.

File: instrumentation/rack/lib/opentelemetry/instrumentation/rack/middlewares/event_handler.rb
Comment (arielvalentin) [reply instrumentation/rack/lib/opentelemetry/instrumentation/rack/middlewares/event_handler.rb:103]: Going to hold off adding other events until I get this initial implementation merged. 

I am also going to stick with `on_finish`, i.e. after the response has been written to the socket and the `Rack::BufferedResponse` is closed.

**Comment by arielvalentin** [comment]
moving back to draft until I am able to determine why tests are failing

**Comment by arielvalentin** [comment]
Tests are failing for combination of rack >= 2.2 and rack test 2.1

**Comment by arielvalentin** [comment]
I have opened an issue with `rack-test` https://github.com/rack/rack-test/issues/335

Hopefully I will be able to dig in something soon and see what is going on.

**Comment by arielvalentin** [comment]
`jeremyevans` set me straight. The test case setup was wrong.

File: instrumentation/rack/lib/opentelemetry/instrumentation/rack/instrumentation.rb
Comment (fbogsany) [inline instrumentation/rack/lib/opentelemetry/instrumentation/rack/instrumentation.rb:34]: What's the long term plan (you may have answered this elsewhere)?

File: instrumentation/rack/lib/opentelemetry/instrumentation/rack/middlewares/tracer_middleware.rb
Comment (fbogsany) [inline instrumentation/rack/lib/opentelemetry/instrumentation/rack/middlewares/tracer_middleware.rb:11]: Any reason not to update this to use the `config[:allowed_rack_request_headers]` and `config[:allowed_rack_response_headers]` precomputed in the `Instrumentation` class?

File: instrumentation/rack/lib/opentelemetry/instrumentation/rack/middlewares/event_handler.rb
Comment (fbogsany) [inline instrumentation/rack/lib/opentelemetry/instrumentation/rack/middlewares/event_handler.rb:237]: Is there any reason not to cache all these config options in instance variables?

File: instrumentation/rack/lib/opentelemetry/instrumentation/rack/instrumentation.rb
Comment (arielvalentin) [reply instrumentation/rack/lib/opentelemetry/instrumentation/rack/instrumentation.rb:34]: This configuration setting and helper will exist until...

1. We drop support for Sinatra versions that require Rack 1.x
2. A future release that makes the Event Handler the default way for users to instrument their applications

File: instrumentation/rack/lib/opentelemetry/instrumentation/rack/middlewares/event_handler.rb
Comment (arielvalentin) [reply instrumentation/rack/lib/opentelemetry/instrumentation/rack/middlewares/event_handler.rb:237]: The existing instrumentation supports being able to inject the middleware and then initializing the SDK afterwards. 

Using instance variables made it so the instance in the middleware had stale values and did not reflect the values after the instrumentation was installed so I didn't have a choice but to look them up.

Comment (robbkidd) [reply instrumentation/rack/lib/opentelemetry/instrumentation/rack/middlewares/event_handler.rb:237]: My understanding is that caching config option values in instance variables is not a blocker to adding the event-based implementation. 👍🏻 

I think caching (I'm assuming "via memoization" here) is something we could consider doing in a future release if/when the middleware-based implementation is _removed_.

File: instrumentation/rack/lib/opentelemetry/instrumentation/rack/instrumentation.rb
Comment (robbkidd) [reply instrumentation/rack/lib/opentelemetry/instrumentation/rack/instrumentation.rb:57]: @fbogsany @robertlaurin Is this a blocker for merging and releasing the event-based implementation?

**Review by robbkidd — APPROVED** [review]
It hangs together, new behavior is behind a feature flag config option, old behavior works in the handful of Racky test apps I have.

```ruby
# in a Gemfile
gem 'opentelemetry-instrumentation-rack', git: 'https://github.com/open-telemetry/opentelemetry-ruby-contrib', glob: 'instrumentation/*/*.gemspec'
```

I'd like to see the instrumentation's options appear documented in the gem's README, but (1) of the _existing_ options, only `url_quantization` is mentioned so this is not really a hard criteria for this new option and (2) I can be the change I want to see in the world (I'll open a PR updating the README).

**Comment by arielvalentin** [comment]
OOOOOOOOPPPPPPPPPPPPSSSSSS I had set this to auto-merge by accident. I hope that is not a problem.

**Pull request merged**
